src/App.tsx

```diff
@@ -1,80 +1,61 @@
 import Login from './components/Login';
 import './App.css';
-import { useContext, useEffect, useState } from 'react';
-import { ExceptionType } from '@perspect3vism/ad4m';
-import { ExceptionInfo } from '@perspect3vism/ad4m/lib/src/runtime/RuntimeResolver';
-import { Loader, Stack } from '@mantine/core';
-import { Ad4mContext } from '.';
+import { useContext, useState } from 'react';
+import { Button, Loader, Stack, TextInput } from '@mantine/core';
 import TrustAgent from './components/TrustAgent';
 import Navigation from './components/Navigation';
+import { Ad4minContext } from './context/Ad4minContext';
 
 const App = () => {
-  const ad4mClient = useContext(Ad4mContext);
+  const {state: {
+    connected, isUnlocked, candidate, connectedLaoding, did
+  }, methods: {
+    handleTrustAgent,
+    setUrl
+  }} = useContext(Ad4minContext);
 
-  const [connected, setConnected] = useState(false);
-  const [isLogined, setIsLogined] = useState<Boolean>(false);
-  const [did, setDid] = useState("");
-  const [candidate, setCandidate] = useState("");
+  const [url, setURL] = useState("");
 
-  useEffect(() => {
-    const checkConnection = async () => {
-      try {
-        await ad4mClient.runtime.hcAgentInfos(); // TODO runtime info is broken
-        console.log("get hc agent infos success.");
-        setConnected(true);
-      } catch (err) {
-        setConnected(false);
-      }
-    }
-
-    checkConnection();
-
-    console.log("Check if ad4m service is connected.")
-  }, [ad4mClient]);
-
-  const handleLogin = (login: Boolean, did: string) => {
-    setIsLogined(login);
-    setDid(did);
-
-    if (login) {
-      ad4mClient.runtime.addExceptionCallback((exception: ExceptionInfo) => {
-        if (exception.type === ExceptionType.AgentIsUntrusted) {
-          setCandidate(exception.addon!);
-        }
-        Notification.requestPermission()
-          .then(response => {
-            if (response === 'granted') {
-              new Notification(exception.title, { body: exception.message })
-            }
-          });
-        console.log(exception);
-        return null
-      })
-    }
-  }
-
-  const handleTrustAgent = (candidate: string) => {
-    setCandidate(candidate);
+  const onUrlChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    event.preventDefault();
+    const { value } = event.target;
+    setURL(value);
   }
 
   return (
     <div className="App">
       {!connected && (
         <Stack align="center" spacing="xl" style={{margin: "auto"}}>
           <Header />
-          <Loader />
+          {connectedLaoding ? (
+              <Loader />
+            ) : (
+              <>
+                <TextInput 
+                  label="Ad4m URL" 
+                  placeholder='https://www.example.com' 
+                  radius="md" 
+                  size="md" 
+                  onChange={onUrlChange}
+                />
+                <Button onClick={() => setUrl(url)}>
```

Comment:
The ad4m-host url should be configured in the settings page, which means when app starts, it still connect to `localhost:4000/graphql`, but can be reset to a new one in settings page.
In this way, user don't need to input the url everytime when run ad4min, also can configure a new endpoint in the settings page. Is it make sense?

Reply:
You don't need to input the URL all the time, once you set it's stored in localStorage. The current flow now to change the URL is first to lock the agent and then there is a button on the login page which opens a modal to change the URL.

Reply:
I've put in `localhost:4000/graphql` without `ws` or `wss` and now get a blank page with this in the console: `Uncaught DOMException: Failed to construct 'WebSocket': The URL's scheme must be either 'ws' or 'wss'. 'localhost' is not allowed.` which persists as well until I prune/change the entry in localStorage.

Reply:
I will add in validation to fix this